sui_programmability/examples/frenemies/sources/frenemies.move

```diff
@@ -0,0 +1,518 @@
+    fun get_outcome(rank_idx: u64, num_validators: u64): u8 {
+        if (rank_idx < num_validators / 3) {
+            FRIEND // friend = top third
+        } else if (rank_idx >= num_validators * 2 / 3) {
+            ENEMY // enemy = bottom third
+        } else {
+            NEUTRAL // neutral = middle third
+        }
+    }
+
+    /// Return the index of the validator with address `addr` in `v`
+    fun find_validator(v: &vector<Validator>, addr: address): u8 {
+        let i = 0;
+        let len = vector::length(v);
+        while (i < len) {
+            let e = vector::borrow(v, i);
+            if (e.addr == addr) {
+                return (i as u8)
+            };
+            i = i + 1
+        };
+        // unreachable with static validator set. TODO: make more robust?
+        abort(0)
+    }
+
+    /// Sort active validators in `state` by stake and add to `leaderboard prev_epoch_stakes` so we can score
+    /// incoming scorecards.
+    fun update_epoch_info(state: &SuiSystemState, leaderboard: &mut Leaderboard, epoch: u64) {
+        if (leaderboard.start_epoch == epoch) {
+            // do not record info for the start epoch, since it is not scored
+            return
+        };
+        let validators = vector[];
+        let active_validators = validator_set::active_validators(sui_system::validators(state));
+        let i = 0;
+        let len = vector::length(active_validators);
+        while (i < len) {
+            let validator_data = vector::borrow(active_validators, i);
+            validator_insertion_sort(
+                &mut validators,
+                Validator {
+                    addr: validator::sui_address(validator_data),
+                    stake: validator::total_stake(validator_data),
+                }
+            );
+            i = i + 1
+        };
+        table::add(&mut leaderboard.prev_epoch_stakes, epoch, validators);
+        leaderboard.epoch = epoch + 1
+    }
+
+    /// insert `validator` into `v`, maintaining the invariant that `v` is in descending order by stake
+    fun validator_insertion_sort(v: &mut vector<Validator>, validator: Validator) {
+        let stake = validator.stake;
+        let i = 0;
+        let len = vector::length(v);
+        while (i < len) {
+            if (stake > vector::borrow(v, i).stake) {
+                vector::insert(v, validator, i);
+                return
+            };
+            i = i + 1
+        };
+        vector::push_back(v, validator)
+    }
+
+     /// write `s` into `v`, maintaining the invariant that `v` is in descending order by (score, participation)
+    fun score_insertion_sort(v: &mut vector<Score>, new_score: Score) {
+        let i = vector::length(v);
+        let score = new_score.score;
+        while (i > 0) {
+            let s = vector::borrow_mut(v, i - 1);
+            if (score > s.score || score == s.score && new_score.participation > s.participation) {
```

Comment:
This should be `new_score.participation > s.participation` right? Since less participation means higher rank

Reply:
Yes, good catch! Fixed + tested much better in latest.